alpenhorn/queue.py

```diff
@@ -0,0 +1,426 @@
+        """
+
+        if wait > 0:
+            # Deferred put
+            with self._dlock:
+                # If joining, silently discard this put
+                if not self._joining:
+                    heapq.heappush(
+                        self._deferrals,
+                        (
+                            time() + wait,
+                            item,
+                            key,
+                        ),
+                    )
+        else:
+            # Immediate put
+            with self._not_empty:
+                self._put(item, key)
+                self._not_empty.notify()  # wakes up a single waiting thread
+
+    def _get(self, t: float) -> tuple[Any, Hashable] | None:
+        """Iterate the `get()` loop once for at most `t` seconds.
+
+        Never call this function directly.  Use get() instead.
+
+        Parameters
+        ----------
+        t : float
+            Maximum timeout (in seconds) for this iteration of the
+            `get()` loop.
+
+        Returns
+        -------
+        If there was nothing to get, returns None.
+
+        Otherwise:
+
+        item : anything
+            The item removed from the queue
+        key : hashable
+            The name of the FIFO from which `item` was popped
+        """
+        timeout_at = time() + t
+        with self._dlock:
+            # If there are delayed puts, don't wait past the expiry of
+            # the earliest
+            if len(self._deferrals) > 0:
+                first_expiry = min(self._deferrals)[0]
+                if timeout_at > first_expiry:
+                    timeout_at = first_expiry
+
+        # Wait until t seconds elapse, or there's something to get
+        wait = timeout_at - time()
+        if wait > 0 and self._total_queued == 0:
+            self._not_empty.wait(wait)
+
+        # Execute all the expired deferred puts
+        with self._dlock:
+            while len(self._deferrals) > 0 and min(self._deferrals)[0] <= time():
+                _, item, key = heapq.heappop(self._deferrals)
+                self._put(item, key)
+
+        # If the queue is still empty, time out
+        if self._total_queued < 1:
+            return None
+
+        # Otherwise, get the next item from the queue:
+
+        # Choose a FIFO by walking _keys_by_inprogress: find the lowest non-empty
+        # set that has a non-empty FIFO in it.
+        key = None
+        for key_set in self._keys_by_inprogress:
+            if len(key_set) > 0:
+                # Pull FIFOs out of the set until we find a non-empty one
+                empties = set()
+                while len(key_set) > 0:
+                    key = key_set.pop()
+                    if len(self._fifos[key]) == 0:
+                        empties.add(key)
+                        key = None
+                    else:
+                        break
+
+                # Put all the empty FIFOs back
+                key_set.update(empties)
```

Comment:
```suggestion
                # Pull FIFOs out of the set until we find a non-empty one
                for k in key_set:
                    if len(self._fifos[k]) > 0:
                        key = k
                        break
```
I think this could just be a for loop which grabs the first non-empty fifo instead of discarding and re-updating with empty queues.